OCaml 5.6 support from https://github.com/ocaml-community/utop/pull/519
(head 64f4486318c128f531f95fddd1f19fa5b0c95191), plus the move of
Misc.Utf8_lexeme to the top-level Utf8_lexeme module in OCaml 5.6.

diff --git a/src/lib/uTop_compat.ml b/src/lib/uTop_compat.ml
index e4879f0..9030bca 100644
--- a/src/lib/uTop_compat.ml
+++ b/src/lib/uTop_compat.ml
@@ -28,7 +28,11 @@ let destruct_ldot p s =
 (* Check whether an identifier is a valid one. *)
 #if OCAML_VERSION >= (5, 3, 0)
 let is_valid_identifier id =
+#if OCAML_VERSION >= (5, 6, 0)
+  Utf8_lexeme.is_valid_identifier id
+#else
   Misc.Utf8_lexeme.is_valid_identifier id
+#endif
 let lax_modname_from_cmi = Unit_info.lax_modname_from_source
 #else
 let is_valid_identifier id =
@@ -185,3 +189,17 @@ let add_cmi_hook f =
     res
   in
   Persistent_env.Persistent_signature.load := load
+
+#if OCAML_VERSION >= (5, 6, 0)
+let lookup_env f x env =
+  try
+    Some (f x env)
+  with Not_found | Env.Error.In_context _ ->
+    None
+#else
+let lookup_env f x env =
+  try
+    Some (f x env)
+  with Not_found | Env.Error _ ->
+    None
+#endif
diff --git a/src/lib/uTop_complete.ml b/src/lib/uTop_complete.ml
index 554f790..11072fe 100644
--- a/src/lib/uTop_complete.ml
+++ b/src/lib/uTop_complete.ml
@@ -25,12 +25,6 @@ let set_of_list = List.fold_left (fun set x -> String_set.add x set) String_set.

 let add id set = if is_valid_identifier id then String_set.add id set else set

-let lookup_env f x env =
-  try
-    Some (f x env)
-  with Not_found | Env.Error _ ->
-    None
-
 (* +-----------------------------------------------------------------+
    | Parsing                                                         |
    +-----------------------------------------------------------------+ *)
diff --git a/src/lib/uTop_lexer.mll b/src/lib/uTop_lexer.mll
index 61b83b0..3789c31 100644
--- a/src/lib/uTop_lexer.mll
+++ b/src/lib/uTop_lexer.mll
@@ -169,6 +169,7 @@ and symbol idx acc = parse
           let loc = mkloc idx (idx + 1) (lexeme_start lexbuf) (lexeme_end lexbuf) in
           tokens loc.idx2 ((Error, loc) :: acc) lexbuf
       }
+  | "" { raise (Failure "lexing: empty token") }

 and cm_string idx= parse
   | '"'
diff --git a/src/lib/uTop_main.ml b/src/lib/uTop_main.ml
index 29bfae5..b2ccde3 100644
--- a/src/lib/uTop_main.ml
+++ b/src/lib/uTop_main.ml
@@ -1481,7 +1481,7 @@ let load_inputrc () =

 let protocol_version = 1

-let main_aux ~initial_env =
+let main_aux ~initial_env : unit =
   Clflags.debug := true;
   Arg.parse args file_argument usage;
 #if OCAML_VERSION >= (5, 0, 0) && OCAML_VERSION < (5, 1, 0)
@@ -1525,7 +1525,7 @@ let main_aux ~initial_env =
   (* Don't let the standard toplevel run... *)
   exit 0

-let main_internal ~initial_env =
+let main_internal ~initial_env : unit =
   try
     main_aux ~initial_env
   with exn ->
@@ -1574,7 +1574,7 @@ let iter_structure expr =
   iter.structure iter

 let interact ?(search_path=[]) ?(build_dir="_build") ~unit ~loc:(fname, lnum, cnum, _)
-      ~values =
+      ~values () =
   let search_path = walk build_dir ~init:search_path ~f:(fun dir acc -> dir :: acc) in
   let cmt_fname =
     try
